# [\#11002 Issue](https://github.com/ethereum/solidity/issues/11002) `open`: msg.data cannot be used inside a receive() function
**Labels**: `language design :rage4:`, `annoys users :cry:`


#### <img src="https://avatars.githubusercontent.com/u/4115788?u=b87c3d4ba9d242ee2a36420b1c7783b848269f7a&v=4" width="50">[sirhill](https://github.com/sirhill) opened issue at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002):

I am migrating from 0.6.12 to 0.8.1 and I am now having this error message:
`CompileError: TypeError: "msg.data" cannot be used inside of "receive" function.`

However, if I move the `msg.data` reference inside a sub function called directly from inside the receive() function it can still work. This is somehow confusing.

I can understand the intent, but all transactions in Ethereum have a `msg.data` even if this one is expected to be null from the receive() context. I do think that it is still relevant to use semantically `msg.data`.

Furthermore, I was unable to find it documented somewhere.

#### <img src="https://avatars.githubusercontent.com/u/13174375?u=52d702cb6bec53b561afa293cf9cd53ef7a63924&v=4" width="50">[hrkrshnn](https://github.com/hrkrshnn) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-785053612):

I find it weird that `msg.data` cannot be accessed inside `receive()`, but it can call an internal function that can access `msg.data`. Even though, `msg.data` would be empty. Maybe we should also disallow it for consistency?

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-785162545):

This is a warning that prevents you from making the invalid assumption that `msg.data` contains useful information inside a `receive` function. As often, most warnings are tailored towards the "regular user" who does not write proxy contracts or anything deeply involved and will hopefully be thankful about such a warning.

#### <img src="https://avatars.githubusercontent.com/u/4115788?u=b87c3d4ba9d242ee2a36420b1c7783b848269f7a&v=4" width="50">[sirhill](https://github.com/sirhill) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-785164767):

It is not a warning but a compilation error.

Your intent was interesting and very valuable.
However, it can also be confusing for a regular user that it can't use `msg.data` (which he will trivially - I think - understand it's null if he is interested to use `receive()` in the first place).

#### <img src="https://avatars.githubusercontent.com/u/4115788?u=b87c3d4ba9d242ee2a36420b1c7783b848269f7a&v=4" width="50">[sirhill](https://github.com/sirhill) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-785168742):

I think that you are also maybe assuming that `msg.data` should not be null and that it should be considered harmful to use it when it's null. Depending of the use case, it can also be a perfectly reasonable value, such as an `if test` checking for the presence or not of such `msg.data`.

#### <img src="https://avatars.githubusercontent.com/u/137030?v=4" width="50">[cameel](https://github.com/cameel) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-785301235):

Maybe we should change it form error to warning then?

#### <img src="https://avatars.githubusercontent.com/u/4115788?u=b87c3d4ba9d242ee2a36420b1c7783b848269f7a&v=4" width="50">[sirhill](https://github.com/sirhill) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-786880106):

If so, I think that maybe, it should rather be a solhint rule as it is quite an opiniated choice.

#### <img src="https://avatars.githubusercontent.com/u/4054346?u=3d0b66be582c6179aadc08d96b8380ad61979fc1&v=4" width="50">[droconnel22](https://github.com/droconnel22) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-810669343):

I am running into this very compiler error* for upgrading a legacy contract from 0.5.14. to 0.8.2. Is it best practice to remove/refactor the usage of `msg.data` or to even refactor from `receive` to `fallback`? I am not comfortable with just changing the solhint error to a warning.

#### <img src="https://avatars.githubusercontent.com/u/4054346?u=3d0b66be582c6179aadc08d96b8380ad61979fc1&v=4" width="50">[droconnel22](https://github.com/droconnel22) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-810671446):

Update: I was able to change the lines:




     receive () payable external {

       // throws solhint error 

       (bool success, ) = myImplContract.delegatecall(msg.data);

       ...


to


    receive () payable external {
        bool success = _mySubfuc();
        ....
    }
 
     function _mySubfuc() private returns(bool) {
          (bool success, ) = myImplContract.delegatecall(msg.data);
          return success;
      }


What is the logic/ justification for allowing this pattern?

#### <img src="https://avatars.githubusercontent.com/u/137030?v=4" width="50">[cameel](https://github.com/cameel) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-811356321):

@droconnel22 `msg.data` is empty in `receive`. Why do you want to pass it to the `delegatecall()`?

If you meant to use a payable `fallback` rather than receive then this is exactly the situation this warning is about. You do not get it when use `msg.data` in a function you call because that function might be called also from places other than `receive` that do have non-empty `msg.data`.

#### <img src="https://avatars.githubusercontent.com/u/2388185?v=4" width="50">[bshastry](https://github.com/bshastry) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-964146494):

This is an error. Perhaps, we can make it a warning?

```
contract C {
  receive() payable external { msg.data; }
}
```

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-964150860):

I would like to see some examples first where you cannot just use an empty calldata byte array instead.

#### <img src="https://avatars.githubusercontent.com/u/20340?v=4" width="50">[axic](https://github.com/axic) commented at [2021-02-24 08:52](https://github.com/ethereum/solidity/issues/11002#issuecomment-964151340):

The `receive` function is supposed to be for non-data payment transactions (i.e. regular eth transfers).

Any other payable function is for received value with `msg.data`.

And for the generic case (proxies, etc.) there is the payable `fallback` function.


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2022.05.23 at 03:51:38.]
